### PR TITLE
First working DQN, wins at trivial game

### PR DESCRIPTION
**Pull request by adamantivm**
This MR includes a few fixes, to the point it was successfully proven to work on a trivial game.

The fixes:
 - Get the reward for the appropriate agent (it was fetching the reward to the opponent before)
 - When playing, initialize epsilon to 0 instead of 1 (it was playing randomly!)
Also a hack:
 - Amplify the reward so that it stands out more clearly from the random initialization weights

Some things are still known to be missing:
 - Training is only on blacks at the moment, we should extend it to whites
 - The negative reward for losing is being lost (not pun intended) and not used for training currently, it should be fixed
 - There are tons of opportunities for improvement in DQN architecture
 - Try with a more realistic game!

But at least this is now a baseline that shows to be working!

If you run the training with the following params:

```
/home/julian/aaae/deep-rabbit-hole/code/deep_rabbit_hole/.venv/bin/python /home/julian/aaae/deep-rabbit-hole/code/deep_rabbit_hole/deep_quoridor/src/train_dqn.py -N 3 -W 0 -e 10000 -u 100
```

It reliably produces an agent which can beat random on blacks 100% of the times:

```
(.venv) ➜  deep_rabbit_hole git:(dq/dqn-training-fixes) /home/julian/aaae/deep-rabbit-hole/code/deep_rabbit_hole/.venv/bin/python /home/julian/aaae/deep-rabbit-hole/code/deep_rabbit_hole/deep_quoridor/src/play.py -t 100 -N 3 -W 0 -p random pretrained01flatdqn -r arenaresults
Loading pre-trained model from /home/julian/aaae/deep-rabbit-hole/code/deep_rabbit_hole/models/dqn_agent_final.pt
...
Arena stats
===========
Board size:  3
Max walls:  0
Step rewards:  False
Total number of games:  100 

+---------------------+---------------------+--------+--------+
|       P1 \ P2       | pretrained01flatdqn | random | Total  |
+---------------------+---------------------+--------+--------+
| pretrained01flatdqn |          -          |   0%   |   0%   |
|        random       |          0%         |   -    |   0%   |
|        ======       |        ======       | ====== | ====== |
|        Total        |          0%         |   0%   |   0%   |
+---------------------+---------------------+--------+--------+
```

(the important number is row 2, column 1)

One really interesting thing to note is how things change when we change the reward multiplier (1000, 100, 1), the number of epochs to train, the epsilon decay and how often to save. Tweaking these values produce much more different training experiences. I recommend playing with them to see it first hand.

### CODE REVIEW

File: deep_quoridor/src/train_dqn.py [hunk 1]
```diff
@@ -73,8 +79,13 @@ def train_dqn(
                 # Execute action
                 game.step(action)
 
-                # Get new state, reward, etc.
-                next_observation, reward, termination, truncation, _ = game.last()
+                # Get the observation and rewards for THIS agent (not the opponent)
```
Comment (adamantivm) [inline deep_quoridor/src/train_dqn.py:82]: @alejandromarcu this part is key, otherwise it was getting the rewards for the opponent instead

File: deep_quoridor/src/train_dqn.py [hunk 2]
```diff
@@ -32,11 +32,8 @@ def train_dqn(
     """
     game = env(board_size=board_size, max_walls=max_walls, step_rewards=step_rewards)
 
-    # Calculate action space size
-    action_size = board_size**2 + ((board_size - 1) ** 2) * 2
-
     # Create the DQN agent
-    dqn_agent = FlatDQNAgent(board_size, action_size)
+    dqn_agent = FlatDQNAgent(board_size, epsilon_decay=0.9999)
```
Comment (adamantivm) [inline deep_quoridor/src/train_dqn.py:36]: @alejandromarcu and this is the other bug you found, this attempts to make it decay a bit slower

Comment (alejandromarcu) [reply deep_quoridor/src/train_dqn.py:36]: nit, remove the line above computing action_size.
Also, I'm wondering if the epsilon decay should happen after episode rather than after training step

**Comment by jonbinney** [comment]
I like the commit message "First ever trivial training that works?". We should make all of your commit messages questions :-)

File: deep_quoridor/src/train_dqn.py [hunk 3]
```diff
@@ -59,11 +62,14 @@ def train_dqn(
             observation, reward, termination, truncation, _ = game.last()
 
             # If the game is over, break the loop
+            # BUG: Missing recording a negative reward for the DQN agent when the opponent wins
             if termination or truncation:
+                if False:
+                    print(f"Winner: {game.get_opponent(agent_name)}. Episode reward: {episode_reward}")
```
Comment (alejandromarcu) [inline deep_quoridor/src/train_dqn.py:68]: We should use a logging framework so we can keep this things as `log.debug` or something and we can turn on and off from the command line

**Review by alejandromarcu — APPROVED** [review]
LGTM

File: deep_quoridor/src/agents/flat_dqn.py [hunk 4]
```diff
@@ -183,7 +183,7 @@ class Pretrained01FlatDQNAgent(FlatDQNAgent):
 
     def __init__(self, board_size, **kwargs):
         super(Pretrained01FlatDQNAgent, self).__init__(board_size)
```
Comment (alejandromarcu) [inline deep_quoridor/src/agents/flat_dqn.py:185]:
```suggestion
        super().__init__(board_size, epsilon=0.0)
```
As I mentioned in chat, epsilon should be 0 when playing, otherwise if you don't specify it, it will be 1 so it will be playing randomly.  Additional nit, I removed the parameters to `super`, that was the format in Python 2.0

**Comment by adamantivm** [comment]
This is now ready to merge. i recommend to check it out, it's exciting!
